cpp/src/arrow/compute/function_internal.h

```diff
@@ -382,9 +382,10 @@ static inline Result<std::shared_ptr<Scalar>> GenericToScalar(std::nullopt_t) {
 }
 
 template <typename T>
-static inline auto GenericToScalar(const std::optional<T>& value)
-    -> Result<decltype(MakeScalar(value.value()))> {
-  return value.has_value() ? MakeScalar(value.value()) : std::make_shared<NullScalar>();
+static inline Result<std::shared_ptr<Scalar>> GenericToScalar(
+    const std::optional<T>& value) {
+  return value.has_value() ? GenericToScalar(value.value())
+                           : std::make_shared<NullScalar>();
```

Comment:
The new change builds and passes CI now, however I can't evaluate it critically enough. @zanmato1984 could you review this further?

Reply:
Yes, I'm looking into the rest of the PR.